### ISSUE TITLE
Support JDK 17

### ISSUE THREAD

**Issue by apupier** [jolokia/jolokia#524]
JDK 17 is the latest LTS (Lon-term Support) released as of today.
it would be nice to support it

**Comment by rhuss** [comment]
Definitely a +1 from my side.

**Comment by varikin** [comment]
I am rolling out JDK 17 and tested our current version, Jolokia 1.6.2. Basic functionality is working of getting JMX over HTTP as JSON. Is there more that I should be testing related to 17 and Jolokia?

**Comment by grgrzybek** [comment]
Tests work on JDK 1.7:
```
[INFO] Reactor Summary:
[INFO] 
[INFO] jolokia-parent ..................................... SUCCESS [  0.191 s]
[INFO] jolokia-test-util .................................. SUCCESS [  1.211 s]
[INFO] jolokia-agent-parent ............................... SUCCESS [  0.005 s]
[INFO] jolokia-core ....................................... SUCCESS [ 20.601 s]
[INFO] jolokia-jsr160 ..................................... SUCCESS [  1.192 s]
[INFO] jolokia-war ........................................ SUCCESS [  0.390 s]
[INFO] jolokia-war-unsecured .............................. SUCCESS [  0.055 s]
[INFO] jolokia-osgi ....................................... SUCCESS [  3.577 s]
[INFO] jolokia-osgi-bundle ................................ SUCCESS [  1.293 s]
[INFO] jolokia-jmx ........................................ SUCCESS [  2.142 s]
[INFO] jolokia-jvm ........................................ SUCCESS [ 28.112 s]
[INFO] jolokia-spring ..................................... SUCCESS [  6.042 s]
[INFO] jolokia-mule ....................................... SUCCESS [  2.434 s]
[INFO] jolokia-it-parent .................................. SUCCESS [  0.004 s]
[INFO] jolokia-it-core .................................... SUCCESS [  0.345 s]
[INFO] jolokia-it-war ..................................... SUCCESS [  0.076 s]
[INFO] jolokia-client-parent .............................. SUCCESS [  0.004 s]
[INFO] jolokia-client-java ................................ SUCCESS [  8.701 s]
[INFO] jolokia-client-javascript .......................... SUCCESS [  0.349 s]
[INFO] jolokia-jmx-adapter ................................ SUCCESS [ 38.016 s]
[INFO] jolokia-kubernetes ................................. SUCCESS [  6.582 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:02 min
[INFO] Finished at: 2023-06-29T13:25:07+02:00
[INFO] Final Memory: 97M/591M
[INFO] ------------------------------------------------------------------------

$ java -version
java version "1.7.0_80"
Java(TM) SE Runtime Environment (build 1.7.0_80-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.80-b11, mixed mode)

$ mvn --version
Apache Maven 3.2.3 (33f8c3e1027c3ddde99d3cdebad2656a31e8fdf4; 2014-08-11T22:58:10+02:00)
Maven home: /data/java/tools/apache-maven-3.2.3
Java version: 1.7.0_80, vendor: Oracle Corporation
Java home: /data/java/javase/jdk1.7.0_80/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.3.8-200.fc38.x86_64", arch: "amd64", family: "unix"
```

On JDK 17:
```
[INFO] Reactor Summary for jolokia-parent 1.7.3-SNAPSHOT:
[INFO] 
[INFO] jolokia-parent ..................................... SUCCESS [  0.198 s]
[INFO] jolokia-test-util .................................. SUCCESS [  1.242 s]
[INFO] jolokia-agent-parent ............................... SUCCESS [  0.005 s]
[INFO] jolokia-core ....................................... SUCCESS [ 18.823 s]
[INFO] jolokia-jsr160 ..................................... SUCCESS [  1.161 s]
[INFO] jolokia-war ........................................ SUCCESS [  0.285 s]
[INFO] jolokia-war-unsecured .............................. SUCCESS [  0.047 s]
[INFO] jolokia-osgi ....................................... SUCCESS [  2.679 s]
[INFO] jolokia-osgi-bundle ................................ SUCCESS [  0.860 s]
[INFO] jolokia-jmx ........................................ SUCCESS [  1.616 s]
[INFO] jolokia-jvm ........................................ SUCCESS [ 25.068 s]
[INFO] jolokia-spring ..................................... SUCCESS [  5.633 s]
[INFO] jolokia-mule ....................................... SUCCESS [  1.983 s]
[INFO] jolokia-it-parent .................................. SUCCESS [  0.004 s]
[INFO] jolokia-it-core .................................... SUCCESS [  0.253 s]
[INFO] jolokia-it-war ..................................... SUCCESS [  0.066 s]
[INFO] jolokia-client-parent .............................. SUCCESS [  0.004 s]
[INFO] jolokia-client-java ................................ SUCCESS [  7.328 s]
[INFO] jolokia-client-javascript .......................... SUCCESS [  0.222 s]
[INFO] jolokia-jmx-adapter ................................ SUCCESS [ 44.931 s]
[INFO] jolokia-kubernetes ................................. SUCCESS [  4.900 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:58 min
[INFO] Finished at: 2023-06-29T13:27:17+02:00
[INFO] ------------------------------------------------------------------------

$ java -version
java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment (build 17.0.7+8-LTS-224)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.7+8-LTS-224, mixed mode, sharing)

$ mvn --version
Apache Maven 3.9.3 (21122926829f1ead511c958d89bd2f672198ae9f)
Maven home: /data/java/tools/apache-maven-3.9.3
Java version: 17.0.7, vendor: Oracle Corporation, runtime: /data/java/javase/jdk-17.0.7
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.3.8-200.fc38.x86_64", arch: "amd64", family: "unix"
```

and on JDK 8:
```
[INFO] Reactor Summary for jolokia-parent 1.7.3-SNAPSHOT:
[INFO] 
[INFO] jolokia-parent ..................................... SUCCESS [  0.189 s]
[INFO] jolokia-test-util .................................. SUCCESS [  1.067 s]
[INFO] jolokia-agent-parent ............................... SUCCESS [  0.006 s]
[INFO] jolokia-core ....................................... SUCCESS [ 17.729 s]
[INFO] jolokia-jsr160 ..................................... SUCCESS [  0.967 s]
[INFO] jolokia-war ........................................ SUCCESS [  0.350 s]
[INFO] jolokia-war-unsecured .............................. SUCCESS [  0.044 s]
[INFO] jolokia-osgi ....................................... SUCCESS [  2.404 s]
[INFO] jolokia-osgi-bundle ................................ SUCCESS [  0.826 s]
[INFO] jolokia-jmx ........................................ SUCCESS [  1.487 s]
[INFO] jolokia-jvm ........................................ SUCCESS [ 24.401 s]
[INFO] jolokia-spring ..................................... SUCCESS [  5.551 s]
[INFO] jolokia-mule ....................................... SUCCESS [  1.900 s]
[INFO] jolokia-it-parent .................................. SUCCESS [  0.003 s]
[INFO] jolokia-it-core .................................... SUCCESS [  0.223 s]
[INFO] jolokia-it-war ..................................... SUCCESS [  0.066 s]
[INFO] jolokia-client-parent .............................. SUCCESS [  0.004 s]
[INFO] jolokia-client-java ................................ SUCCESS [  7.323 s]
[INFO] jolokia-client-javascript .......................... SUCCESS [  0.240 s]
[INFO] jolokia-jmx-adapter ................................ SUCCESS [ 39.892 s]
[INFO] jolokia-kubernetes ................................. SUCCESS [  4.889 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:50 min
[INFO] Finished at: 2023-06-29T13:29:43+02:00
[INFO] ------------------------------------------------------------------------

$ java -version
openjdk version "1.8.0_372"
OpenJDK Runtime Environment (build 1.8.0_372-b07)
OpenJDK 64-Bit Server VM (build 25.372-b07, mixed mode)

$ mvn --version
Apache Maven 3.9.3 (21122926829f1ead511c958d89bd2f672198ae9f)
Maven home: /data/java/tools/apache-maven-3.9.3
Java version: 1.8.0_372, vendor: everfree, runtime: /data/java/javase/jdk-8u372/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.3.8-200.fc38.x86_64", arch: "amd64", family: "unix"
```

**Comment by apupier** [comment]
Note: some tests related to Certificates were deactivated https://github.com/jolokia/jolokia/commit/174317b7454f88fae79a7bb883f8e92c2be8197a

**Comment by apupier** [comment]
Note: some tests related to Certificates were deactivated https://github.com/jolokia/jolokia/commit/174317b7454f88fae79a7bb883f8e92c2be8197a https://github.com/jolokia/jolokia/issues/523

**Comment by grgrzybek** [comment]
Works even on JDK 1.6 (after removing `jolokia-kubernetes` module which needs JDK 1.7+):
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] jolokia-parent ..................................... SUCCESS [  0.193 s]
[INFO] jolokia-test-util .................................. SUCCESS [  1.499 s]
[INFO] jolokia-agent-parent ............................... SUCCESS [  0.006 s]
[INFO] jolokia-core ....................................... SUCCESS [ 21.865 s]
[INFO] jolokia-jsr160 ..................................... SUCCESS [  1.099 s]
[INFO] jolokia-war ........................................ SUCCESS [  0.469 s]
[INFO] jolokia-war-unsecured .............................. SUCCESS [  0.084 s]
[INFO] jolokia-osgi ....................................... SUCCESS [  3.296 s]
[INFO] jolokia-osgi-bundle ................................ SUCCESS [  8.009 s]
[INFO] jolokia-jmx ........................................ SUCCESS [  1.354 s]
[INFO] jolokia-jvm ........................................ SUCCESS [ 26.321 s]
[INFO] jolokia-spring ..................................... SUCCESS [  6.141 s]
[INFO] jolokia-it-parent .................................. SUCCESS [  0.004 s]
[INFO] jolokia-it-core .................................... SUCCESS [  0.265 s]
[INFO] jolokia-it-war ..................................... SUCCESS [  0.109 s]
[INFO] jolokia-client-parent .............................. SUCCESS [  0.005 s]
[INFO] jolokia-client-java ................................ SUCCESS [  8.316 s]
[INFO] jolokia-client-javascript .......................... SUCCESS [  0.358 s]
[INFO] jolokia-jmx-adapter ................................ SUCCESS [ 27.996 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:48 min
[INFO] Finished at: 2023-06-29T14:52:54+02:00
[INFO] Final Memory: 39M/421M
[INFO] ------------------------------------------------------------------------

$ java -version
java version "1.6.0_45"
Java(TM) SE Runtime Environment (build 1.6.0_45-b06)
Java HotSpot(TM) 64-Bit Server VM (build 20.45-b01, mixed mode)

$ mvn --version
Apache Maven 3.2.3 (33f8c3e1027c3ddde99d3cdebad2656a31e8fdf4; 2014-08-11T22:58:10+02:00)
Maven home: /data/java/tools/apache-maven-3.2.3
Java version: 1.6.0_45, vendor: Sun Microsystems Inc.
Java home: /data/java/javase/jdk1.6.0_45/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.3.8-200.fc38.x86_64", arch: "amd64", family: "unix"
```

**Comment by grgrzybek** [comment]
For completeness, JDK 11:
```
[INFO] Reactor Summary for jolokia-parent 1.7.3-SNAPSHOT:
[INFO] 
[INFO] jolokia-parent ..................................... SUCCESS [  0.188 s]
[INFO] jolokia-test-util .................................. SUCCESS [  1.243 s]
[INFO] jolokia-agent-parent ............................... SUCCESS [  0.006 s]
[INFO] jolokia-core ....................................... SUCCESS [ 18.677 s]
[INFO] jolokia-jsr160 ..................................... SUCCESS [  1.348 s]
[INFO] jolokia-war ........................................ SUCCESS [  0.278 s]
[INFO] jolokia-war-unsecured .............................. SUCCESS [  0.048 s]
[INFO] jolokia-osgi ....................................... SUCCESS [  2.821 s]
[INFO] jolokia-osgi-bundle ................................ SUCCESS [  0.899 s]
[INFO] jolokia-jmx ........................................ SUCCESS [  1.832 s]
[INFO] jolokia-jvm ........................................ SUCCESS [ 28.255 s]
[INFO] jolokia-spring ..................................... SUCCESS [  5.904 s]
[INFO] jolokia-mule ....................................... SUCCESS [  2.185 s]
[INFO] jolokia-it-parent .................................. SUCCESS [  0.004 s]
[INFO] jolokia-it-core .................................... SUCCESS [  0.285 s]
[INFO] jolokia-it-war ..................................... SUCCESS [  0.063 s]
[INFO] jolokia-client-parent .............................. SUCCESS [  0.004 s]
[INFO] jolokia-client-java ................................ SUCCESS [  7.841 s]
[INFO] jolokia-client-javascript .......................... SUCCESS [  0.231 s]
[INFO] jolokia-jmx-adapter ................................ SUCCESS [ 43.562 s]
[INFO] jolokia-kubernetes ................................. SUCCESS [  5.072 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:01 min
[INFO] Finished at: 2023-06-29T15:13:52+02:00
[INFO] ------------------------------------------------------------------------

$ java -version
java version "11.0.19" 2023-04-18 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.19+9-LTS-224)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.19+9-LTS-224, mixed mode)

$ mvn --version
Apache Maven 3.9.3 (21122926829f1ead511c958d89bd2f672198ae9f)
Maven home: /data/java/tools/apache-maven-3.9.3
Java version: 11.0.19, vendor: Oracle Corporation, runtime: /data/java/javase/jdk-11.0.19
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.3.8-200.fc38.x86_64", arch: "amd64", family: "unix"
```

**Comment by grgrzybek** [comment]
thanks @apupier for info - I didn't know it.
BTW I have some interesting FIPS-related scenarios which cause problems with some algorithms. I'm working on it.

**Comment by grgrzybek** [comment]
@apupier @tadayosi @rhuss @varikin I don't think anything more has to be done here.

**Comment by tadayosi** [comment]
@grgrzybek I agree. If I could add one more, I noticed that the CI doesn't assert support for 17 yet.
https://github.com/jolokia/jolokia/blob/master/.github/workflows/ci.yaml#L17
If we want to clarify we support the specific versions, maybe it's better to add all of them to the workflow.

**Comment by grgrzybek** [comment]
I know very little about the workflows, so feel free to change it ;)

**Comment by tadayosi** [comment]
Let's close. Fixed with #547.
